### PR TITLE
[Feature] Fast forward server to latest changes

### PR DESCRIPTION
**Pull request by SferaDev**
### :pushpin: References

* **Issue:** Closes https://github.com/EyeSeeTea/metadata-synchronization/issues/440

### :memo: Implementation

- Fast-forward logic from version 0.4.0 to 1.0.0
- Add support for ``aggregated`` and ``events`` sync rules aside from ``metadata``
- Adapt code to use ``d2-api``, ``migrations`` and latest changes to ``dataStore``
- Improve logger messages

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -23,47 +27,60 @@ app.use("/", indexRouter);
...
+    .options({
+        config: {
+            type: "string",
+            alias: "c",
+            describe: "Configuration file",
+            default: path.join(__dirname, rootFolder, "app-config.json"),
+        },
+    })
+    .coerce("config", path => {
+        if (fs.existsSync(path)) {
```
Comment (tokland) [inline src/index.ts]: Minor: Cant we simply run read+parse and let the functions throw the exceptions?

File: src/index.ts [hunk 2]
```diff
@@ -23,47 +27,60 @@ app.use("/", indexRouter);
...
-}).argv;
+};
 
 const start = async (): Promise<void> => {
-    const { encryptionKey, apiUrl: baseUrl, username, password } = config;
+    const { encryptionKey, baseUrl, username, password } = config;
+
+    const authorization = `Basic ${btoa(username + ":" + password)}`;
+    const api = new D2ApiDefault({ baseUrl, auth: { username, password } });
+    const d2 = await init({ baseUrl: `${baseUrl}/api`, headers: { authorization } });
```
Comment (tokland) [inline src/index.ts:72]: Ideally, we should provide all this d2models info (AFAICS, it's the only thing we use `d2` for), in d2-api in the future, right?

File: src/logic/scheduler.ts [hunk 3]
```diff
@@ -1,67 +1,99 @@
...
 import SyncRule from "../models/syncRule";
+import { D2 } from "../types/d2";
+import { SynchronizationRule, SyncRuleType } from "../types/synchronization";
+import { AggregatedSync } from "./sync/aggregated";
+import { DeletedSync } from "./sync/deleted";
+import { EventsSync } from "./sync/events";
+import { SyncronizationClass } from "./sync/generic";
+import { MetadataSync } from "./sync/metadata";
+
+const config: Record<
```
Comment (tokland) [inline src/logic/scheduler.ts]: [minor] I would write this type as in a one-liner, it looks more clear.

File: src/logic/scheduler.ts [hunk 4]
```diff
@@ -1,67 +1,99 @@
...
+        const rule = await SyncRule.get(this.api, id);
+        const { name, frequency, builder, id: syncRule, type = "metadata" } = rule;
+        const { SyncClass } = config[type];
 
-    private synchronizationTask = async (syncRule: SynchronizationRule): Promise<void> => {
-        const { id, name, builder, frequency } = syncRule;
         const logger = getLogger(name);
         try {
-            logger.debug(`Start with frequency: ${cronstrue.toString(frequency || "")}`);
-            for await (const { message, syncReport, done } of startSynchronization(this.d2, {
```
Comment (tokland) [inline src/logic/scheduler.ts:23]: For future projects/iterations: there is quite some replicated code app/server. 

@SferaDev  Do you think it made sense to have a split project or it's better to keep them together? We can discuss it some time. One option to keep it in separate repositories would be adding one project as a dependency of the other.

File: src/models/d2ModelFactory.ts [hunk 5]
```diff
@@ -1,41 +1,56 @@
...
+    classes.ValidationRuleGroupModel,
+    classes.ProgramModel,
+    classes.ProgramIndicatorModel,
+    classes.ProgramIndicatorGroupModel,
+    classes.ProgramRuleModel,
+    classes.ProgramRuleVariableModel,
+    classes.UserGroupModel,
+];
+
+const findClasses = (...args: any[]) => {
```
Comment (tokland) [inline src/models/d2ModelFactory.ts]: It's used passing an args, not varadic args. Shouldn't it be `const findClasses = (args: string[]) => {`?

File: src/types/d2.d.ts [hunk 6]
```diff
@@ -17,35 +17,14 @@ export interface Params {
     page?: number;
     pageSize?: number;
     filter?: string[];
-    fields?: string[];
+    fields?: (string | number)[];
```
Comment (tokland) [inline src/types/d2.d.ts]: Where do we use a `number` as fields?

File: README.md [hunk 7]
```diff
@@ -27,7 +27,7 @@ File `app-config.json` must be provided. If no configuration file is supplied th
 ```json
 {
     "encryptionKey": "encryptionKey",
-    "apiUrl": "https://play.dhis2.org/2.30/api",
+    "baseUrl": "https://play.dhis2.org/2.30",
```
Comment (tokland) [inline README.md:30]: If no baseUrl is defined, it continues anyway and tries a `GET undefined/api/schemas`. Maybe we should validate the values from the json?

File: src/utils/dataStore.ts [hunk 8]
```diff
@@ -0,0 +1,22 @@
+import { D2Api } from "d2-api";
+import { dataStoreNamespace, dataStoreVersion } from "../models/dataStore";
+
+type Maybe<T> = T | undefined | null;
+
+export const initializeDataStoreMigrations = async (api: D2Api) => {
```
Comment (tokland) [inline src/utils/dataStore.ts]: I think we don't need this anymore, it's all in src/migrations

**Review by tokland — COMMENTED** [review]
Tested in dev2:8085 (3 sync rules, with scheduling, for aggregated/events/metadata), I had no problems getting the info and running those sync in the server. Some comments about the code:

File: src/index.ts (hunk 2 above)
Comment (SferaDev) [reply src/index.ts:72]: Agreed, it's the only thing we use ``d2`` both in the app and here. Once we have the schema definitions in ``d2-api`` we can remove the dependency altogether.

File: src/logic/scheduler.ts (hunk 4 above)
Comment (SferaDev) [reply src/logic/scheduler.ts:23]: Yes, I will discuss this @adrianq in the next meeting.

**Comment by adrianq** [comment]
@SferaDev Can I close this in favour of https://github.com/EyeSeeTea/metadata-synchronization/pull/497 ? I guess you include @tokland 's comments into that other PR, right?

**Comment by SferaDev** [comment]
> @SferaDev Can I close this in favour of [EyeSeeTea/metadata-synchronization#497](https://github.com/EyeSeeTea/metadata-synchronization/pull/497) ? I guess you include @tokland 's comments into that other PR, right?

Yes, all the requested changes are in both repos

**Pull request closed without merging**